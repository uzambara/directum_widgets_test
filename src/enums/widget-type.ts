export enum WidgetType {
    Weather = 1,
    CurrencyRate = 2
}