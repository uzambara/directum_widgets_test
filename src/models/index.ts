import {INamedEntityWithId} from "./shared/named-entity-with-id";
import {ICity} from "./shared/city";
import {ICurrency} from "./currency-widget/currency";
import {IWidget} from "./shared/widget";
import {ICurrencyRateWidget} from "./currency-widget/currency-rate-widget";
import {IWeatherWidget} from "./weather-widget/weather-widget";
import {ICurrencyRateWidgetData} from "./currency-widget/currency-rate-widget-data";
import {IWeatherWidgetData} from "./weather-widget/weather-widget-data";

export {
    ICity,
    ICurrency,
    INamedEntityWithId,
    IWidget,
    ICurrencyRateWidget,
    IWeatherWidget,
    ICurrencyRateWidgetData,
    IWeatherWidgetData
}