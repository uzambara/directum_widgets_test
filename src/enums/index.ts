import {WidgetType} from "./widget-type";
import {WeatherType} from "./weather/weather-type";
import {PrecipitationType} from "./weather/precipitation-type"
import {CurrencyType} from "./currency/currency-type";
import {TemperatureUnitType} from "./weather/temperature-unit-type";


export {
    WidgetType,
    WeatherType,
    PrecipitationType,
    TemperatureUnitType,
    CurrencyType
};