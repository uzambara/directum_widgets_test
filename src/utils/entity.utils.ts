import {INamedEntityWithId} from "../models";

const createSelectorByKey = <T extends INamedEntityWithId>(entities: Array<T>) => (key: any) => {
    let filteredByKey = entities.filter(entity => entity.id === key);
    if(filteredByKey.length > 0) {
        return filteredByKey[0];
    }
    return null;
};

export const entityUtils = {
    createSelectorByKey
};