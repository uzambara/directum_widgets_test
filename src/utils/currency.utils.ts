import {CurrencyType} from "../enums";
import {currencyImages} from "../images";

function getCurrencyImageByType(currencyType: CurrencyType): string {
    switch (currencyType) {
        case CurrencyType.EUR:
            return currencyImages.eur;
        case CurrencyType.USD:
            return currencyImages.usd;
        case CurrencyType.GBP:
            return currencyImages.gbp;
        case CurrencyType.UAH:
            return currencyImages.uah;
        default:
            return "";
    }
}

export const currencyUtils = {
    getCurrencyImageByType
};