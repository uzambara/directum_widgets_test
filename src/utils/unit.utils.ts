import {TemperatureUnitType} from "../enums";

function getFarengatesFromCelsius(celsius: number) {
    return parseFloat((celsius * 9 / 5 + 32).toFixed(2));
}

function getTemperatureByUnitType(celsiusTemp: number, unitType: TemperatureUnitType) {
    switch (unitType) {
        case TemperatureUnitType.Fahrenheit:
            return getFarengatesFromCelsius(celsiusTemp).toFixed(0);
        case TemperatureUnitType.Celsius:
            return celsiusTemp.toFixed(2);
        default:
            return "unknown";
    }
}

function getTemperatureUnitHtml(temperatureUnit: TemperatureUnitType) {
    switch (temperatureUnit) {
        case TemperatureUnitType.Celsius:
            return "&#8451;";
        case TemperatureUnitType.Fahrenheit:
            return "&#8457;";
        default:
            return "unknown";
    }
}

export const unitUtils = {
    getFarengatesFromCelsius,
    getTemperatureUnitHtml,
    getTemperatureByUnitType
};