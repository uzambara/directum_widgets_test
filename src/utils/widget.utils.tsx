import * as React from "react";
import {ICurrencyRateWidget, IWeatherWidget, IWidget} from "../models";
import {WidgetType} from "../enums";
import {ValidationResult, Validator} from "lakmus";
import {CurrencyRateWidgetValidator, WeatherWidgetValidator} from "../validators";
import {MutableRefObject} from "react";
import _ from "underscore";

function initWidgetsByColumnMap(columnsCount: number) {
    let result = {};
    _.range(columnsCount).map((idx) => result[idx] = []);

    return result;
}

type GetNewRowIndexAfterDragStopParams = {
    currentColumnElementRef: MutableRefObject<HTMLDivElement>,
    currentRowIndex: number,
    currentColumnIndex: number,
    newColumnIdx: number,
    changedWidgetTop: number
}

function getNewRowIndexAfterDragStop({
        currentColumnElementRef,
        currentRowIndex,
        currentColumnIndex,
        newColumnIdx,
        changedWidgetTop}: GetNewRowIndexAfterDragStopParams): number {
    const {current} = currentColumnElementRef;
    let newRowIndex = 0;
    const isTheSameColumn = currentColumnIndex == newColumnIdx;
    for(; newRowIndex < current.children.length; newRowIndex++) {
        if(newRowIndex == currentRowIndex && isTheSameColumn)
            continue;
        const currentWidgetRect = current.children[newRowIndex].getBoundingClientRect();
        if (changedWidgetTop < currentWidgetRect.top) {
            if(newRowIndex - 1 == currentRowIndex && isTheSameColumn) {
                newRowIndex--;
            }
            break;
        }
    }
    return newRowIndex;
}

type GetNewColumnIndexAfterDragStopParams = {
    columnWidth: number,
    columnsCount: number,
    currentColumnIndex: number,
    lastX: number

}

function getNewColumnIndexAfterDragStop({
        columnWidth,
        columnsCount,
        currentColumnIndex,
        lastX}: GetNewColumnIndexAfterDragStopParams): number {
    const halfWidthOfColumn = columnWidth / 2;
    const x = Math.abs(lastX);
    const halfWidthOfColumnCountOffset = Math.floor(x / halfWidthOfColumn);
    let move = 0;

    if(halfWidthOfColumnCountOffset > 0) {
        move = (Math.floor(halfWidthOfColumnCountOffset / 2) + halfWidthOfColumnCountOffset % 2) * Math.sign(lastX);
    }

    let newColumnIdx = currentColumnIndex + move;

    if(newColumnIdx < 0 || newColumnIdx > columnsCount - 1) {
        newColumnIdx = currentColumnIndex;
    }
    return newColumnIdx;
}

function deleteWidget(widget: IWidget, widgetsByColumn: {[columnIdx: number]: IWidget[]}): {[columnIdx: number]: IWidget[]} {
    const {columnIndex, id} = widget;
    const filtered = widgetsByColumn[columnIndex].filter(_widget => _widget.id !== id);
    return {
        ...widgetsByColumn,
        [columnIndex]: filtered
    }
}

function changeWidget(newWidget: IWidget, oldWidget: IWidget, widgetsByColumn: {[columnIdx: number]: IWidget[]}): {[columnIdx: number]: IWidget[]} {
    newWidget.initialData = null;
    if(newWidget.columnIndex === oldWidget.columnIndex) {
        const newWidgets = widgetsByColumn[oldWidget.columnIndex].map(_widget => _widget.id === oldWidget.id ? newWidget : _widget);
        const result = {
            ...widgetsByColumn,
            [oldWidget.columnIndex]: newWidgets
        };
        return result;
    }

    const oldColumnWidgets = widgetsByColumn[oldWidget.columnIndex].filter(_widget => _widget.id !== oldWidget.id);
    const newColumnWidgets = widgetsByColumn[newWidget.columnIndex] || [];
    newColumnWidgets.push(newWidget);
    return {
        ...widgetsByColumn,
        [oldWidget.columnIndex]: oldColumnWidgets,
        [newWidget.columnIndex]: newColumnWidgets
    }
}

function addWidget(widget: IWidget, widgetsByColumn: {[columnIdx: number]: IWidget[]}) {
    let widgetsWithSameColumn = widgetsByColumn[widget.columnIndex];
    if(!widgetsWithSameColumn) {
        widgetsWithSameColumn = [];
    }
    widgetsWithSameColumn.push(widget);
    return {
        ...widgetsByColumn,
        [widget.columnIndex]: widgetsWithSameColumn
    };
}

function changeWidgetPlace(
        id: string,
        columnIndex: number,
        newColumnIndex: number,
        newRowIndex: number,
        widgetsByColumn: {[columnIdx: number]: IWidget[]}) {
    const widgetForChanges = widgetsByColumn[columnIndex].find(w => w.id === id) as IWeatherWidget;
    widgetForChanges.columnIndex = newColumnIndex;
    const currentRowIndex = widgetsByColumn[columnIndex].indexOf(widgetForChanges);

    if(columnIndex == newColumnIndex) {
        if(currentRowIndex == newRowIndex) {
            return widgetsByColumn;
        }

        const filteredArr = widgetsByColumn[newColumnIndex].filter(w => w.id !== id);
        filteredArr.splice(newRowIndex, 0, widgetForChanges);

        return {
            ...widgetsByColumn,
            [newColumnIndex]: filteredArr
        };
    }
    let widgetsCopy = [...widgetsByColumn[newColumnIndex] || []];
    widgetsCopy.splice(newRowIndex, 0, widgetForChanges);
    return {
        ...widgetsByColumn,
        [columnIndex]: widgetsByColumn[columnIndex].filter(w => w.id !== id),
        [newColumnIndex]: widgetsCopy
    };
}

function validateWidget(widget: IWidget): ValidationResult {
    let validator: Validator<IWeatherWidget | ICurrencyRateWidget> = null;
    let result: ValidationResult = null;
    switch(widget.type) {
        case WidgetType.CurrencyRate:
            validator = new CurrencyRateWidgetValidator();
            result = validator.validate(widget as ICurrencyRateWidget);
            break;
        case WidgetType.Weather:
            validator = new WeatherWidgetValidator();
            result = validator.validate(widget as IWeatherWidget);
            break;
    }
    return result;
}

export const widgetUtils = {
    validateWidget,
    getNewColumnIndexAfterDragStop,
    getNewRowIndexAfterDragStop,
    deleteWidget,
    changeWidgetPlace,
    addWidget,
    changeWidget,
    initWidgetsByColumnMap
};