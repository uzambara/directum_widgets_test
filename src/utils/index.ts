import {widgetUtils} from "./widget.utils"
import {unitUtils} from "./unit.utils";
import {fetchUtil} from "./fetch.utils";
import {currencyUtils} from "./currency.utils";
import {guid} from "./guid.utils";

export {
    widgetUtils,
    unitUtils,
    fetchUtil,
    currencyUtils,
    guid
}