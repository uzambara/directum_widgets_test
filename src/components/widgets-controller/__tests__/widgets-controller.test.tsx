import {act} from "react-dom/test-utils";

jest.mock("../../../utils/guid.utils");
import {shallow} from "enzyme";
import React from "react";
import {CurrencyRateWidgetSettings, WeatherWidgetSettings, WidgetsController} from "../../index";
import {ICity, IWeatherWidget} from "../../../models";
import renderer from "react-test-renderer";
import {TemperatureUnitType, WidgetType} from "../../../enums";
import {WidgetsControllerComponent} from "../../index";
import {widgetUtils} from "../../../utils";
import {ValidationResult} from "lakmus";
import {Alert} from "@material-ui/lab";

const CITIES: ICity[] = [
    {id: "id1", displayName: "city1", lon: "lon1", lat: "lat1"},
    {id: "id2", displayName: "city2", lon: "lon2", lat: "lat2"},
    {id: "id3", displayName: "city3", lon: "lon2", lat: "lat3"}
];

const onSubmitMock = jest.fn();
const onCancelMock = jest.fn();
const className = "className";

describe("widgets controller tests", () => {
    beforeEach(() => {
        jest.resetAllMocks();
    });

    it("Render", () => {
        // Arrange
        const columnsCount = 2;
        let submitButtonText = "OK";
        let initialWidget: IWeatherWidget = {
            city: CITIES[0],
            columnIndex: 0,
            id: "123",
            temperatureUnit: TemperatureUnitType.Fahrenheit,
            type: WidgetType.Weather

        };

        // Act
        const snapshot = renderer.create(<WidgetsController
            cities={CITIES}
            columnsCount={columnsCount}
            onSubmit={onSubmitMock}
            onCancel={onCancelMock}
            className={className}
            submitButtonText={submitButtonText}
            initialWidget={initialWidget}
        />).toJSON();

        // Asserts
        expect(snapshot).toMatchSnapshot();
    });

    it("Меняет значение колонки", async () => {
        // Arrange
        const columnsCount = 2;
        const component = shallow(<WidgetsControllerComponent
            cities={CITIES}
            columnsCount={columnsCount}
            onSubmit={onSubmitMock}
        />);

        // Act
        let columnNumberSelect = component.find('[labelText="Column number"]');
        act(() => {
            columnNumberSelect.props().onChange({target: {value: 999}} as any);
        });

        // Asserts
        columnNumberSelect = component.find('[labelText="Column number"]');
        expect(columnNumberSelect.props().value).toBe(999);
    });

    it("Меняет значение типа виджета", async () => {
        // Arrange
        const columnsCount = 2;
        const component = shallow(<WidgetsControllerComponent
            cities={CITIES}
            columnsCount={columnsCount}
            onSubmit={onSubmitMock}
        />);

        // Act
        let widgetTypeSelect = component.find('[labelText="Widget type"]');
        act(() => {
            widgetTypeSelect.props().onChange({target: {value: 888}} as any);
        });

        // Asserts
        widgetTypeSelect = component.find('[labelText="Widget type"]');
        expect(widgetTypeSelect.props().value).toBe(888);
    });

    it("Рендерит нужный контрол в зависимости от типа виджета", async () => {
        // Arrange
        const columnsCount = 2;
        const component = shallow(<WidgetsControllerComponent
            cities={CITIES}
            columnsCount={columnsCount}
            onSubmit={onSubmitMock}
        />);

        // Act
        let widgetTypeSelect = component.find('[labelText="Widget type"]');
        act(() => {
            widgetTypeSelect.props().onChange({target: {value: -1}} as any);
        });

        // Asserts
        widgetTypeSelect = component.find('[labelText="Widget type"]');
        expect(widgetTypeSelect.props().value).toBe(-1);
        expect(component.find(CurrencyRateWidgetSettings)).toHaveLength(0);
        expect(component.find(WeatherWidgetSettings)).toHaveLength(0);


        // Act
        widgetTypeSelect = component.find('[labelText="Widget type"]');
        act(() => {
            widgetTypeSelect.props().onChange({target: {value: WidgetType.Weather}} as any);
        });

        // Asserts
        widgetTypeSelect = component.find('[labelText="Widget type"]');
        expect(widgetTypeSelect.props().value).toBe(WidgetType.Weather);
        expect(component.find(CurrencyRateWidgetSettings)).toHaveLength(0);
        expect(component.find(WeatherWidgetSettings)).toHaveLength(1);


        // Act
        widgetTypeSelect = component.find('[labelText="Widget type"]');
        act(() => {
            widgetTypeSelect.props().onChange({target: {value: WidgetType.CurrencyRate}} as any);
        });

        // Asserts
        widgetTypeSelect = component.find('[labelText="Widget type"]');
        expect(widgetTypeSelect.props().value).toBe(WidgetType.CurrencyRate);
        expect(component.find(CurrencyRateWidgetSettings)).toHaveLength(1);
        expect(component.find(WeatherWidgetSettings)).toHaveLength(0);
    });

    it("Рендерит кнопку Cancel, если передан обоработчик. Вызывает обработчик при нажатии на нее.", () => {
        // Arrange
        const columnsCount = 2;
        const component = shallow(<WidgetsControllerComponent
            cities={CITIES}
            columnsCount={columnsCount}
            onSubmit={onSubmitMock}
            onCancel={onCancelMock}
        />);

        // Act
        const cancelButton = component.find('[text="Cancel"]');
        expect(cancelButton).toHaveLength(1);
        act(() => {
            cancelButton.props().onClick(null);
        });

        // Asserts
        expect(onCancelMock).toHaveBeenCalled();
    });

    it("Рендерит кнопку Submit. Вызывает обработчик при нажатии на нее если прошла валидация.", () => {
        // Arrange
        const columnsCount = 2;
        const component = shallow(<WidgetsControllerComponent
            cities={CITIES}
            columnsCount={columnsCount}
            onSubmit={onSubmitMock}
            onCancel={onCancelMock}
            submitButtonText="Submit"
        />);

        // Act
        const submitButton = component.find('[text="Submit"]');
        expect(submitButton).toHaveLength(1);
        widgetUtils.validateWidget = jest.fn().mockImplementation((): ValidationResult => ({
            isValid: true,
            errors: null
        }));
        act(() => {
            submitButton.props().onClick(null);
        });

        // Asserts
        expect(onSubmitMock).toHaveBeenCalled();
    });

    it("Рендерит ошибку и не вызывает обработчик при нажатии на нее если валидация не прошла.", () => {
        // Arrange
        const columnsCount = 2;
        const component = shallow(<WidgetsControllerComponent
            cities={CITIES}
            columnsCount={columnsCount}
            onSubmit={onSubmitMock}
            onCancel={onCancelMock}
            submitButtonText="Submit"
        />);

        // Act
        const submitButton = component.find('[text="Submit"]');
        expect(submitButton).toHaveLength(1);
        widgetUtils.validateWidget = jest.fn().mockImplementation((): ValidationResult => ({
            isValid: false,
            errors: [{} as any]
        }));
        act(() => {
            submitButton.props().onClick(null);
        });

        // Asserts
        expect(onSubmitMock).not.toHaveBeenCalled();
        const errorAlert = component.find(Alert);
        expect(errorAlert).toHaveLength(1);
        expect(errorAlert.text()).toBe("Need to choose all fields.");
    });
});