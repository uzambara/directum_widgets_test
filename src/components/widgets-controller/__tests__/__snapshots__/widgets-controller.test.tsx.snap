// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`widgets controller tests Render 1`] = `
<section
  className="className"
>
  <div
    className="MuiFormControl-root"
  >
    <label
      className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-marginDense MuiFormLabel-filled"
      data-shrink={true}
    >
      Column number
    </label>
    <div
      className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl MuiInputBase-marginDense MuiInput-marginDense"
      onClick={[Function]}
    >
      <div
        aria-haspopup="listbox"
        className="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiInputBase-input MuiInput-input MuiInputBase-inputMarginDense MuiInput-inputMarginDense"
        onBlur={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onMouseDown={[Function]}
        role="button"
        tabIndex={0}
      >
        1
      </div>
      <input
        onAnimationStart={[Function]}
        type="hidden"
        value={0}
      />
      <svg
        aria-hidden="true"
        className="MuiSvgIcon-root MuiSelect-icon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M7 10l5 5 5-5z"
        />
      </svg>
    </div>
  </div>
  <div
    className="MuiFormControl-root"
  >
    <label
      className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-marginDense MuiFormLabel-filled"
      data-shrink={true}
    >
      Widget type
    </label>
    <div
      className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl MuiInputBase-marginDense MuiInput-marginDense"
      onClick={[Function]}
    >
      <div
        aria-haspopup="listbox"
        className="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiInputBase-input MuiInput-input MuiInputBase-inputMarginDense MuiInput-inputMarginDense"
        onBlur={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onMouseDown={[Function]}
        role="button"
        tabIndex={0}
      >
        Weather
      </div>
      <input
        onAnimationStart={[Function]}
        type="hidden"
        value={1}
      />
      <svg
        aria-hidden="true"
        className="MuiSvgIcon-root MuiSelect-icon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M7 10l5 5 5-5z"
        />
      </svg>
    </div>
  </div>
  <div
    className=""
  >
    <div
      className="MuiFormControl-root"
    >
      <label
        className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-marginDense MuiFormLabel-filled"
        data-shrink={true}
      >
        City
      </label>
      <div
        className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl MuiInputBase-marginDense MuiInput-marginDense"
        onClick={[Function]}
      >
        <div
          aria-haspopup="listbox"
          className="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiInputBase-input MuiInput-input MuiInputBase-inputMarginDense MuiInput-inputMarginDense"
          onBlur={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onMouseDown={[Function]}
          role="button"
          tabIndex={0}
        >
          city1
        </div>
        <input
          onAnimationStart={[Function]}
          type="hidden"
          value="id1"
        />
        <svg
          aria-hidden="true"
          className="MuiSvgIcon-root MuiSelect-icon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M7 10l5 5 5-5z"
          />
        </svg>
      </div>
    </div>
    <div
      className="MuiFormControl-root"
    >
      <label
        className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-marginDense MuiFormLabel-filled"
        data-shrink={true}
      >
        Temp. unit
      </label>
      <div
        className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl MuiInputBase-marginDense MuiInput-marginDense"
        onClick={[Function]}
      >
        <div
          aria-haspopup="listbox"
          className="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiInputBase-input MuiInput-input MuiInputBase-inputMarginDense MuiInput-inputMarginDense"
          onBlur={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onMouseDown={[Function]}
          role="button"
          tabIndex={0}
        >
          <span
            dangerouslySetInnerHTML={
              Object {
                "__html": "&#8457;",
              }
            }
          />
        </div>
        <input
          onAnimationStart={[Function]}
          type="hidden"
          value={2}
        />
        <svg
          aria-hidden="true"
          className="MuiSvgIcon-root MuiSelect-icon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M7 10l5 5 5-5z"
          />
        </svg>
      </div>
    </div>
  </div>
  <div>
    <div>
      <button
        className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        <span
          className="MuiButton-label"
        >
          OK
        </span>
      </button>
      <button
        className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary"
        disabled={false}
        onBlur={[Function]}
        onClick={[MockFunction]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        <span
          className="MuiButton-label"
        >
          Cancel
        </span>
      </button>
    </div>
  </div>
</section>
`;

exports[`widgets controller tests render 1`] = `
<section
  className="className"
>
  <div
    className="MuiFormControl-root"
  >
    <label
      className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-marginDense MuiFormLabel-filled"
      data-shrink={true}
    >
      Column number
    </label>
    <div
      className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl MuiInputBase-marginDense MuiInput-marginDense"
      onClick={[Function]}
    >
      <div
        aria-haspopup="listbox"
        className="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiInputBase-input MuiInput-input MuiInputBase-inputMarginDense MuiInput-inputMarginDense"
        onBlur={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onMouseDown={[Function]}
        role="button"
        tabIndex={0}
      >
        1
      </div>
      <input
        onAnimationStart={[Function]}
        type="hidden"
        value={0}
      />
      <svg
        aria-hidden="true"
        className="MuiSvgIcon-root MuiSelect-icon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M7 10l5 5 5-5z"
        />
      </svg>
    </div>
  </div>
  <div
    className="MuiFormControl-root"
  >
    <label
      className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-marginDense MuiFormLabel-filled"
      data-shrink={true}
    >
      Widget type
    </label>
    <div
      className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl MuiInputBase-marginDense MuiInput-marginDense"
      onClick={[Function]}
    >
      <div
        aria-haspopup="listbox"
        className="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiInputBase-input MuiInput-input MuiInputBase-inputMarginDense MuiInput-inputMarginDense"
        onBlur={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onMouseDown={[Function]}
        role="button"
        tabIndex={0}
      >
        Weather
      </div>
      <input
        onAnimationStart={[Function]}
        type="hidden"
        value={1}
      />
      <svg
        aria-hidden="true"
        className="MuiSvgIcon-root MuiSelect-icon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M7 10l5 5 5-5z"
        />
      </svg>
    </div>
  </div>
  <div
    className=""
  >
    <div
      className="MuiFormControl-root"
    >
      <label
        className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-marginDense MuiFormLabel-filled"
        data-shrink={true}
      >
        City
      </label>
      <div
        className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl MuiInputBase-marginDense MuiInput-marginDense"
        onClick={[Function]}
      >
        <div
          aria-haspopup="listbox"
          className="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiInputBase-input MuiInput-input MuiInputBase-inputMarginDense MuiInput-inputMarginDense"
          onBlur={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onMouseDown={[Function]}
          role="button"
          tabIndex={0}
        >
          city1
        </div>
        <input
          onAnimationStart={[Function]}
          type="hidden"
          value="id1"
        />
        <svg
          aria-hidden="true"
          className="MuiSvgIcon-root MuiSelect-icon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M7 10l5 5 5-5z"
          />
        </svg>
      </div>
    </div>
    <div
      className="MuiFormControl-root"
    >
      <label
        className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-marginDense MuiFormLabel-filled"
        data-shrink={true}
      >
        Temp. unit
      </label>
      <div
        className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl MuiInputBase-marginDense MuiInput-marginDense"
        onClick={[Function]}
      >
        <div
          aria-haspopup="listbox"
          className="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiInputBase-input MuiInput-input MuiInputBase-inputMarginDense MuiInput-inputMarginDense"
          onBlur={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onMouseDown={[Function]}
          role="button"
          tabIndex={0}
        >
          <span
            dangerouslySetInnerHTML={
              Object {
                "__html": "&#8457;",
              }
            }
          />
        </div>
        <input
          onAnimationStart={[Function]}
          type="hidden"
          value={2}
        />
        <svg
          aria-hidden="true"
          className="MuiSvgIcon-root MuiSelect-icon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M7 10l5 5 5-5z"
          />
        </svg>
      </div>
    </div>
  </div>
  <div>
    <div>
      <button
        className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        <span
          className="MuiButton-label"
        >
          OK
        </span>
      </button>
      <button
        className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary"
        disabled={false}
        onBlur={[Function]}
        onClick={[MockFunction]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        <span
          className="MuiButton-label"
        >
          Cancel
        </span>
      </button>
    </div>
  </div>
</section>
`;
